### ISSUE TITLE
onnxruntime slower than xgboost & lightgbm in batch predictions

### ISSUE THREAD

**Issue by AbhilashAwasthi01** [microsoft/onnxruntime#886]
Have a xgboost and a lightgbm classifier model. 
Converted both to ONNX Model format and then onnx model read using onnxruntime 
onnxruntime (onnx model) for batch prediction is approx 5 to 10 times slower than the respective xgboost/lightgbm models in case of batch predictions.
 
Took batch size of 150 for predictions and with 21 numeric features.

Runtimes (in milliseconds)

|XGBoost(n_estimators=500)|ONNXRuntime|OS|
|:---------:|:-----:|:-----:|
|1.7 | 13.5 | macOS mojave|
|1.77| 25 | Linux Ubuntu 16.04|

|LightGBM(n_estimators=1000) |ONNXRuntime|OS|
|:---------:|:-----:|:-----:|
|4.94 | 22.5 | macOS mojave|
|1.77| 25 | Linux Ubuntu 16.04|

**System information**
- OS Platform and Distribution: Tested both on Mac and Ubuntu 16.04.5
- ONNX Runtime installed from: binary (using pip)
- ONNX Runtime version: 0.3.0
- Python version: 3.6

**Comment by faxu** [comment]
Thanks for flagging this. It appears related to https://github.com/Microsoft/onnxruntime/issues/827 
It's on our backlog to investigate further, but we also welcome contributions if you'd like to take a look.

**Comment by xadupre** [comment]
xgboost is multithreaded, onnxruntime is not for ensembles. That might explain the differences. How many cores does your CPU have?

**Comment by AbhilashAwasthi01** [comment]
Okay thanks for clearing this that onnxruntime is not for ensembles.
My Mac has 8 cores and ubuntu machine has 24 cores.

**Comment by RandySheriffH** [comment]
@AbhilashAwasthi01: would u mind share related script?
Will reproduce the issue and profile the model.

**Comment by RandySheriffH** [comment]
Close the issue by fix - https://github.com/microsoft/onnxruntime/pull/1276

**Comment by snnn** [comment]
Reopen it because the fix got reverted

**Comment by faxu** [comment]
@xadupre Can this be closed with the latest round of improvements in 1.3?

**Comment by giorgostheo** [comment]
I can see that this issue is bit old, but I still find that ONNX is much much slower than vanilla XGB even in low-core machines like a raspberry pi... Is this something that should/will be resolved eventually or is it out of the question?

**Comment by xadupre** [comment]
Sorry, I missed this a couple of years ago. We currently have fixed settings for parallelization but they should depend on the processor (cache size) and the problem size (tree depth, number of features). It is difficult to take the right decision in the generic case but we could expose the parallelization settings. You would have to optimize on your own to find the best settings but it is possible to make the computation 2 to 10 times faster.

**Comment by giorgostheo** [comment]
Thanks for the quick reply! Sounds great, is there a pull request that implements this or a branch or smth? I would gladly test this on my platforms.

**Comment by xadupre** [comment]
Not yet. I'm still wondering about I should implement it.

**Comment by xadupre** [comment]
I can expose some parameters in onnxruntime config but the same parameter would be the same for every tree in the forest or use field description in the node to be able to set these parameters in a different way for every forest in the model.

**Comment by giorgostheo** [comment]
I can see (from the PR discussion as well) that the main issue is finding out which param set is best. If you like (and can expose them somehow) I can potentially run tests in multiple machines and architectures to find out if there is way to approximate it. I am still a bit of a noob though tho :)

**Comment by xadupre** [comment]
That would be great. Let me a day to prepare a PR and an example.

**Comment by xadupre** [comment]
This is something I did a while ago. I just move onnxruntime implementation (the old one) in a custom kernel and I tried many different parallelization settings: [TreeEnsemble optimization](https://sdpython.github.io/doc/onnx-extended/dev/auto_examples/plot_op_tree_ensemble_optim.html). This script should work but I cannot say this would compile on ARM. I just did it on CPU. I modified the onnxruntime implementation based on those results. The only conclusion I can make is the same settings does not work well on all processors and all datasets. The random computation is like a matrix computation, in a grid, every thread is computing k trees with n observations, k < T total trees and n < N total number of observations. How to choose efficiently k and n?

**Comment by giorgostheo** [comment]
Just FYI, performance is not that bad for onnx (certainly not 10x slower that the issue suggests, it is mostly better actually!). Its just that picking N and K that @xadupre mentions can really impact performance, which can be huge - especially in low-powered devices.  Since we all like numbers, here are some for TreeEnsembleClassifier and different batch sizes on an M2Max and Rpi 5 (time is for the whole batch not per sample, N=number of iterations to get the avg):

M2MAX

XGB | BS=1, N=100 -> Time: 9.117 ms
XGB | BS=10, N=100 -> Time: 9.372 ms
XGB | BS=100, N=100 -> Time: 9.777 ms
XGB | BS=1000, N=100 -> Time: 9.685 ms
XGB | BS=10000, N=100 -> Time: 10.122 ms

ONNX | BS=1, N=100 -> Time: 8.096 ms
ONNX | BS=10, N=100 -> Time: 7.178 ms
ONNX | BS=100, N=100 -> Time: 7.377 ms
ONNX | BS=1000, N=100 -> Time: 7.537 ms
ONNX | BS=10000, N=100 -> Time: 9.460 ms

Rpi5 16GB

XGB | BS=1, N=100 -> Time: 32.889 ms
XGB | BS=10, N=100 -> Time: 33.692 ms
XGB | BS=100, N=100 -> Time: 34.034 ms
XGB | BS=1000, N=100 -> Time: 34.326 ms
XGB | BS=10000, N=100 -> Time: 34.567 ms

ONNX | BS=1, N=100 -> Time: 31.532 ms
ONNX | BS=10, N=100 -> Time: 32.342 ms
ONNX | BS=100, N=100 -> Time: 32.602 ms
ONNX | BS=1000, N=100 -> Time: 32.995 ms
ONNX | BS=10000, N=100 -> Time: 36.345 ms (!!!)

Hope this helps somehow :)

**Comment by xadupre** [comment]
You probably hit a threshold and the strategy to parallelize changed for something worse.

**Comment by xadupre** [comment]
I created a PR #24147 which lets you modify the parallelization. You need to modify the model to put the parameter inside the model in field ``nodes_hitrates_as_tensor``. The PR description includes a function doing it. The meaning of them is defined here: https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h#L38.

I compiled with the following command line:

```bash
python ./tools/ci_build/build.py --config Release --build_wheel --build_dir ./build/linux_cpu --build_shared_lib --parallel --skip_tests --allow_running_as_root
```

You should be able to call the function I made inside your script before onnxruntime loads the model.

**Comment by giorgostheo** [comment]
PR appears to be working. Currently (in my initial model that was transformed from XGB), "parallel_tree" has by far the most impact of the 3 params. Default param must be something small (<10 I think, I haven't searched though), because base onnx needs about 0.26 ms for 1 sample whereas with parallel_tree of >60 inference drops very quickly to 0.035ms. This is the case of an x86 machine with an AMD 7900X.

Should I create multiple models with different sizes using something like [this](https://sdpython.github.io/doc/onnx-extended/dev/auto_examples/plot_op_tree_ensemble_optim.html) to test? What's our goal here? 

Anyhow, exposing those params to the user seems to be very important to reduce inference times overall :)

**Comment by xadupre** [comment]
The goal for me is to answer two questions:

* Should we export the parallelization parameters? Based on your feedback, it is yes. And then, which mechanism do we use? This one or another. This one allows the user to have different settings for different tree nodes in the model. But I guess settings in SessionOptions would eeasier.
* Should we define default values depending on the processor (at compile time)? We could then set up the values you found efficient in your case as the default for your processor.

**Comment by giorgostheo** [comment]
IMO having different params for different nodes seems redundant. It is rather hard to picture such a case, with the only exception being very imbalanced trees, but still... Session options seem more intuitive and "well placed". Just my 2 cents though...

Regarding the defaults, i guess the main variables at play are the core count and maybe cache size. Even then, the options seem to be endless. On the other hand, if you think more about architectures, maybe different defaults for arm64 work better, but I think it will still be mostly because arm64 processors are generally weaker than x86 etc.

Another idea that might be worth considering is that it might be not too expensive computationally to have some sort of tuning script (at compile or at runtime - I am not sure whether that could work or not) with a fixed range for the 3 variables (or some of them even) that runs and identifies the defaults dynamically. I would bet that having larger values for the 3 params will make inference faster generally, but maybe you will see some diminishing returns eventually...

**Comment by giorgostheo** [comment]
Quick addition. Performance is quite wonky for different batch size in the same CPU. See the best and worst params for different sizes bellow (BS is batch size, (pt, ptn, pn) are the params and time is ms). Each metric is from 100 runs. 

<img width="586" alt="Image" src="https://github.com/user-attachments/assets/e51bd227-635d-4e44-92a1-da6a3ec02eb4" />

**Comment by giorgostheo** [comment]
Results for RPi 5 16GB. Seems more consistent overall, but there are still gains to be had by having better param sets.

<img width="687" alt="Image" src="https://github.com/user-attachments/assets/f6fb490b-31be-462c-a5f0-f79ff8d62be0" />
